### DISCUSSION TITLE
CacheBench bandwidth reporting

### DISCUSSION THREAD

**Discussion by pbhandar2** [facebook/CacheLib#57] [Q&A] [answered]
I am replaying a few KV traces using CacheBench. I had a few questions about the throughput numbers reported in the result. 

- Is the throughput affected by cache misses? Since there is no actual data transfer would a cache miss be simulated by taking into account the time taken to get the data of a cache miss from persistent storage (possible hard drive) to the cache? 
- Can we configure the persistent storage to simulate the cache miss in any way?

**Comment by sathyaphoenix** [comment]
@pbhandar2  

> Is the throughput affected by cache misses? Since there is no actual data transfer would a cache miss be simulated by taking into account the time taken to get the data of a cache miss from persistent storage (possible hard drive) to the cache?

When a cache miss occurs, cachebench by default does not do anything to fill the missing key. However, if you enable the "lookAside" option in the config(see [this](https://cachelib.org/docs/Cache_Library_User_Guides/Configuring_cachebench_parameters#operation-ratios)),  then CacheBench will insert a new value for the missing key. This is similar to how cachebench inserts keys when you configure the setRatio. With this, the cache miss will impact the throughput since the cache will see more sets. However, it does not incorporate the time taken for the cache to fetch the data from backend. To do that, you'll need add some logic in CacheBench to emulate that accordingly. 

> Can we configure the persistent storage to simulate the cache miss in any way?

not sure what you mean by this.  Are you referring to the storage that you pass to cachebench for the hybrid cache ? If so, how do you plan to use that to emulate a cache miss ?

**Reply by pbhandar2** [reply]
Thank you for your response. I was looking to incorporate data transfer from persistent storage on a cache miss like it would when used as a block cache or even other KV solution that are not entirely in-memory. Right now, it seems that cachelib sets the value of the missing key if lookaside option is enabled. 

What I meant by configuring the persistent storage was also in a scenario where the main data lies in persistent storage (probably HDD) and the HybridCache functions as just a cache.

**Comment by sathyaphoenix** [comment, marked as answer]
@pbhandar2 you can hook into the [Stressor framework] (https://github.com/facebook/CacheLib/blob/5cbbfaae0c74e88ab5003e338ff74a0ee75d063d/cachelib/cachebench/runner/Stressor.h) to write your own stressor that reads the value from some other source (that incurs the bandwidth).

**Reply by pbhandar2** [reply]
@sathyaphoenix Thank you so much for the help. I hooked into Cache Stressor to make I/O request to the backend storage. I wanted to confirm one another thing. Currently, the stressor does not replay traces based on the timestamp even if the timestamp field is set in the CacheBench Request object, right? Time based replay is something that has to be implemented separately?
